clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1284,14 +1334,22 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     // Don't do -1 here because we count on this to be the first parameter added
     // (if any).
     size_t ParamIndex = Params.size();
-    for (const ParmVarDecl *Param : InitMethod->parameters())
-      addParam(BS, Param->getType().getCanonicalType());
+    auto ParamIt = InitMethod->parameters().begin();
+    if (*ParamIt) {
+      // Add meaningful argument (not '-1') to buffer_location attribute only
+      // for an accessor pointer
+      size_t BufferLocAttrArg = handleBufferLocationProperty(FieldTy);
+      addParam(BS, (*ParamIt)->getType().getCanonicalType(), BufferLocAttrArg);
+      ++ParamIt;
+      for (; ParamIt != InitMethod->parameters().end(); ++ParamIt)
+        addParam(BS, (*ParamIt)->getType().getCanonicalType(), -1);
+    }
```

Comment:
Feels like I've seen this before. Can we unify it?

Reply:
Do you mean `handleAccessorType` and `handleSyclSpecialType` functions? :)

I have slightly rewrote this part (now I'm obtaining source location, and paths for FD and BS are different).